docker/Makefile

```diff
@@ -0,0 +1,104 @@
+	@NODE_VERSION=$$(node --version | sed 's/v//'); \
+	if [ "$$(echo $$NODE_VERSION $(NODE_VERSION_MIN) | tr " " "\n" | sort -V | head -n1)" != "$(NODE_VERSION_MIN)" ]; then \
+		echo "Node.js version $$NODE_VERSION is too low. Please update to v$(NODE_VERSION_MIN)."; exit 1; \
+	fi
+
+check-yarn:
+	@command -v yarn >/dev/null 2>&1 || { echo >&2 "Yarn is not installed. Please install Yarn v$(YARN_VERSION_MIN) or higher."; exit 1; }
+	@YARN_VERSION=$$(yarn --version); \
+	if [ "$$(echo $$YARN_VERSION $(YARN_VERSION_MIN) | tr " " "\n" | sort -V | head -n1)" != "$(YARN_VERSION_MIN)" ]; then \
+		echo "Yarn version $$YARN_VERSION is too low. Please update to v$(YARN_VERSION_MIN)."; exit 1; \
+	fi
+
+check-rust:
+	@command -v rustc >/dev/null 2>&1 || { echo >&2 "Rust is not installed. Please install Rust v$(RUST_VERSION)."; exit 1; }
+	@command rustup install $(RUST_VERSION) >/dev/null 2>&1
+
+check-sqlx-cli:
+	@command -v sqlx >/dev/null 2>&1 || { echo >&2 "sqlx-cli is not installed. Please install sqlx-cli v$(SQLX_CLI_VERSION)"; exit 1; }
+	@SQLX_CLI_VERSION_LOCAL=$$(sqlx --version | cut -d' ' -f2); \
+	if [ "$$(echo $$SQLX_CLI_VERSION_LOCAL $(SQLX_CLI_VERSION) | tr " " "\n" | sort -V | head -n1)" != "$(SQLX_CLI_VERSION)" ]; then \
+		echo "sqlx-cli version $$SQLX_CLI_VERSION is wrong. Please update to v$(SQLX_CLI_VERSION)"; exit 1; \
+	fi
+
+check-docker:
+	@command -v docker >/dev/null 2>&1 || { echo >&2 "Docker is not installed. Please install Docker v$(DOCKER_VERSION_MIN) or higher."; exit 1; }
+	@DOCKER_VERSION=$$(docker --version | grep -oE '[0-9]+\.[0-9]+\.[0-9]+'); \
+	if [ "$$(echo $$DOCKER_VERSION $(DOCKER_VERSION_MIN) | tr " " "\n" | sort -V | head -n1)" != "$(DOCKER_VERSION_MIN)" ]; then \
+		echo "Docker version $$DOCKER_VERSION is too low. Please update to v$(DOCKER_VERSION_MIN) or higher."; exit 1; \
+	fi
+
+check-foundry:
+	@command -v forge --version >/dev/null 2>&1 || { echo >&2 "Foundry is not installed. Please install Foundry"; exit 1; }
+	@FORGE_VERSION=$$(forge --version | grep -oE '[0-9]+\.[0-9]+\.[0-9]+'); \
+	if [ "$$(echo $$FORGE_VERSION $(FORGE_MIN_VERSION) | tr " " "\n" | sort -V | head -n1)" != "$(FORGE_MIN_VERSION)" ]; then \
+		echo "Forge version $$FORGE_VERSION is too low. Please update to v$(FORGE_MIN_VERSION) or higher."; exit 1; \
+	fi
+
+# Check for all required tools
+check-tools: check-nodejs check-yarn check-rust check-sqlx-cli check-docker check-foundry
+	@echo "All required tools are installed."
+
+# Check that contracts are checkout properly
+check-contracts:
+	@if [ ! -d ../contracts/l1-contracts/lib/openzeppelin-contracts-upgradeable/lib/erc4626-tests ] || [ -z "$$(ls -A ../contracts/l1-contracts/lib/openzeppelin-contracts-upgradeable/lib/erc4626-tests)" ]; then \
+		echo "l1-contracts git submodule is missing. Please re-download repo with `git clone --recurse-submodules https://github.com/matter-labs/zksync-era.git`"; \
```

Comment:
But you also can run `git submodule update --init --recursive`. 
If you don't mess up the repo it self it will init it properly. 


And i'm not sure that checking completely random file inside contracts is a good idea, it could disappear any time

Reply:
That's true, but I can't find a way to check if the submodule is checkout and presented =(